chia/full_node/mempool_manager.py

```diff
@@ -142,35 +143,82 @@ async def create_bundle_from_mempool(
+                for coin_id in eligible_coin_ids:
+                    coin_spend_in_bundle = next(cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == coin_id)
+                    solution_hash_in_bundle = coin_spend_in_bundle.solution.get_tree_hash()
+                    # See if we processed an item with this coin before
+                    if coin_id in dedup_coin_spends:
+                        # See if the solution was identical
+                        processed_solution_hash, duplicate_cost = dedup_coin_spends[coin_id]
+                        if processed_solution_hash == solution_hash_in_bundle:
+                            # It was, so let's add this to the list of spends to deduplicate
+                            spends_to_dedup.append(coin_spend_in_bundle)
+                            cost_saving += duplicate_cost
+                        else:
+                            # It wasn't, so let's skip this whole transaction
+                            consider_this_item = False
+                            cost_saving = 0
+                            break
+                    else:
+                        # We didn't process an item with this coin before
+                        # If we end up including this transaction, add this item to
+                        # the ones we need to run in isolation and store the cost of
+                        spends_to_run_for_cost.append((coin_id, coin_spend_in_bundle))
+                if consider_this_item:
+                    # Run the spends that we're processing for the first time, for cost
+                    new_dedup_coin_spends: Dict[bytes32, Tuple[bytes32, uint64]] = {}
+                    for coin_id, coin_spend in spends_to_run_for_cost:
+                        spend_cost = uint64(coin_spend.puzzle_reveal.run_with_cost(item.cost, coin_spend.solution)[0])
+                        new_dedup_coin_spends[coin_id] = (coin_spend.solution.get_tree_hash(), spend_cost)
+                        cost_saving += spend_cost
+                    item_cost = item.cost - cost_saving
+                    log.info(f"Cumulative cost: {cost_sum}, fee per cost: {item.fee / item_cost}")
+                    if (
+                        item_cost + cost_sum <= self.limit_factor * self.constants.MAX_BLOCK_COST_CLVM
+                        and item.fee + fee_sum <= self.constants.MAX_COIN_AMOUNT
+                    ):
+                        # Update the deduplicated coin spends map
+                        dedup_coin_spends.update(new_dedup_coin_spends)
+                        # Add all the spends in the transaction except the duplicated ones
+                        for cs in item.spend_bundle.coin_spends:
+                            if cs not in spends_to_dedup:
+                                coin_spends.append(cs)
+                                removals.append(cs.coin)
+                        additions.update(item.additions)
```

Comment:
do you mutate `item.additions`? it seems unintuitive to have removed duplicates from this object.
This looks wrong. It looks like you include the duplicate spends here that have been removed.